### ISSUE TITLE
VAMPNet manage multiple trajectories

### ISSUE THREAD

**Issue by pl992** [deeptime-ml/deeptime#133]
I had a similar discussion with you some days ago and the problem was quite well solved. For a better understanding and for a more precise usage I need to ask you something else regarding managing VAMPNet with multiple trajectories. As long as I understood it isn't important to provide a trajectory but rather a set of instantaneous and lagged frames. This raise a couple of questions:
1. In the documentation the validation data are not shuffled: `loader_val = DataLoader(val_data, batch_size=len(val_data), shuffle=False)` is it necessary for correctness or for just performance purposes?  If what's important is that the tuple (cur_frame,lagged_frame) is consistent then I could shuffle without losing this identity and the training should be the same
2. In this way I could also concatenate many trajectories as long as I keep the identity between current frame and lagged frame. For example:
```
class Data:
    def __init__(self,trjs,lagtime):
        self.data_inst = np.concatenate([trj[:-lagtime] for trj in trjs])
        self.data_lagged = np.concatenate([trj[lagtime:] for trj in trjs])
    def __getitem__(self,item):
        return (self.data_inst[item],self.data_lagged[item])
    def __iter__(self):
        for X,Y in zip(self.data_inst,self.data_lagged):
            yield (X,Y)
```
This class could be what I pass to the pytorch DataLoader and I should be training the network correctly. Is this correct or am I missing something? Moreover this concatenated frames could be shuffled as well

Thank you very much for your help again, and congratulations again for this great library!
**EDIT**
The Data class doesn't work with the DataLoader pytorch I wanted to put there just as an example

**Comment by clonker** [comment]
Hey, these are good questions!

1. For validation data it is not necessary to shuffle. Shuffling's purpose is introducing diffusivity into the gradient during backpropagation and to avoid introducing learning artifacts due to repeating patterns in the data.
2. This is not the right way to go about it because it introduces artificial transitions. This will lead to biased results. Small example: First trajectory is `[0, 0, ...., 0]`, second trajectory is `[1, ...., 1]`. If you just concatenate, you will find transitions `0 -> 1` which are not present according to data.

**Comment by pl992** [comment]
thank you! but the transitions aren't just detected by the tuple (frame,lagged_frame)? so let's say I have 2 trajectories. I first create the tuple (frame,lagged_frame) for the first trajectory ([0,0,...0]) obtaining: 
(0,0),(0,0)....
then I create the tuples for the second trajectories separately: (1,1),(1,1)....
Finally I concatenate the two trajectories tuples obtaining (0,0),(0,0),...(0,0),(1,1),...,(1,1). So if for the training is important just the couple of frame,lagged_frame then in this way I'm not building artificial transitions

**Comment by clonker** [comment]
Ah right, I didnt read your pseudocode correctly. 🙂 It is indeed the case that only individual tuples of `(frame -> timelagged frame)` are used to detect transitions. So what you propose there should work fine!
